### ISSUE TITLE
[Password Managers] Some password manager icons do not show up on input fields

### ISSUE THREAD

**Issue by rianadon** [home-assistant/frontend#10274]
### Checklist

- [X] I have updated to the latest available Home Assistant version.
- [X] I have cleared the cache of my browser.
- [X] I have tried a different browser to see if it is related to my browser.

### Describe the issue you are experiencing

This issue is for discussing remaining password manager bugs that have arisen in spite of now having a polyfill in place to support password managers.

I'm ﻿continuing discussion from #3133, now that we have an issue specific to password manager quirks.

# LastPass

I can reproduce @Nowaker's issue with LastPass. Their browser extension made the interesting design choice of adding their clickable icon to the *background* of the input element.
<img width="444" alt="Screen Shot 2021-10-14 at 8 19 36 PM" src="https://user-images.githubusercontent.com/18223213/137427813-7b09874e-9a1d-489e-aa90-2e91ae5a843d.png">
It looks like there are also mouse move & click listeners somewhere that swap the image when you hover over the LastPass icon and know when you click on it.

LastPass successfully finds our polyfilled elements and adjust their background color (although the mouse move detection does not appear to work), but since they are hidden **the LastPass icon will never be visible**.

# Safari

@soundstorm: Safari adds its password fill menu directly to the `input` element. This menu looks like it is the only way to save a password from the Home Assistant page. However, **I see no other way to making this menu visible other than displaying the polyfill element**.

The silver lining is that once a password is created within Safari, the autofocus lets you fill it in right away.

<img width="563" alt="image" src="https://user-images.githubusercontent.com/18223213/137428468-ffac31d7-75d0-4e82-9922-43003376181f.png">

*Note: it is possible to enter the credentials via Safari preferences, but this is a more roundabout way than using the menu.*

# Other password managers

@creack I tried the Bitwarden extension in Firefox and filling in the password works from the icon next to the address bar. It doesn't look like their extension adds icons like LastPass does, so I'm curious what is not working for you.

The only password manager for which the icon showing up does work is KeePassXC (last tested at the time I wrote the polyfill)

# Solutions

I don't think there is any change we can make that will allow Safari and LastPass to show their password autocomplete icons. If anyone has a burning need to write a PR to fix this issue, I suggest adding a "my password manager doesn't work" button that un-hides the polyfilled elements.

**Comment by creack** [comment]
I didn't try the latest version, it may be already fixed. Bitwarden extension in Chrome simply wouldn't fill the fields on both the web ui and the mobile app.

**Comment by deviantintegral** [comment]
Something is up with filling on iOS too. Earlier today, I couldn't get autofill working in Safari at all. I'm using 1Password, but normally use the system-level filling so it should be the same behaviour for any password manager. Then, it worked, and I added the app to my home screen, but when logging in again (as session data is separated) there's no offer to fill at all.

**Comment by leonelsr** [comment]
Google Chrome's own password manager (without extensions) also doesn't work. The same happens on Chromium-based Microsoft Edge, vanilla password manager doesn't recognize the fields.

**Comment by Nowaker** [comment]
# SOLUTION: Stop using fancy JavaScript shadow DOM frameworks to build a freaking login page. A simple HTML from the 90s will do better!

**Comment by leonelsr** [comment]
> SOLUTION: Stop using fancy JavaScript shadow DOM / single page application / whatever comes after Web 2.0 frameworks to build a freaking login page. A simple HTML from the 90s will do better!

Ok, I understand that putting such a request in `h1` (or markdown's `#`) letters doesn't seem to be a very polite approach 😅 Also nobody will simply stop using frameworks and stuff. But, **ffs**, he does have a point!

At least, let's please care having proper and working **polyfills** (also called _"shims"_) always in place, in order to cover most current use cases and scenarios.

I'm not even talking about backwards compatibility (as in IE11 down to IE6... whatsoever), but password managers are increasingly becoming popular these days, as digital security is such a big concern for everyone.

They (even "vanilla" ones) make it easier for users to generate, set and use passwords that are unique for each service, app or website. But usually those are very hard to remember OR type, and ain't nobody likes having to copy and paste stuff anywhere!!!

All that to say this should be treated as a serious, big and important UX bug, because **it is!!!** 😬

Thanks! Best Regards.

**Comment by Nowaker** [comment]
> Ok, I understand that putting such a request in `h1` (or markdown's `#`) letters doesn't seem to be a very polite approach

Before I answer that, let's look at the background first:

> Oct 23, 2019 - **8 upvotes** - @Nowaker - I guess the login page should be a regular HTML login page if SPA makes the basics overly complicated.

> Sep 24, 2020 - **29 upvotes** - @Nowaker - Can we simply stop using the latest tech that is password filler and usability unfriendly and go back to normal HTML? We don't need the 2020 tech on the login page. 1995 tech sans the gradients and shadows will do...

>  Apr 18, 2021 - **19 upvotes** - @Nowaker - The login page should be a regular HTML login page if SPA makes the basics overly complicated.

> Apr 18, 2021 - **30 upvotes ** - @johanson - The founder seems to suggest it wont' be fixed - #1622. I agree with @Nowaker, common sense would be the login page to work with password managers, not aiming for to be 100% x framework. 

It's been almost three years since #3133 was reported, and two years since not using SPA for the login page was suggested, and still counting.

Secondly, password manager support ticket (#3133) is the TOP1 most upvoted ticket of all times for HA frontend - 172 upvotes, ticket is closed, and the problems are still arising. The second most upvoted ticket is #4614 with only 37 upvotes.

So, commenting on what you said - of course, it wasn't polite, but closing and locking the TOP1 most upvoted ticket wasn't polite either. The use of h1 is just a tribute to these 172 upvotes, since the maintainers do not consider "use 90s tech on login page for simplicity" a valid request, as evidenced in #1622.

**Comment by rianadon** [comment]
I don't like moderating here but I would like to clear some things up before the discussion becomes off-topic.

> Also nobody will simply stop using frameworks and stuff. But, ffs, he does have a point!

This is why the login page is not a simple page. Using the same framework & components as the rest of HA makes the page easier to maintain. And the login page isn't always just a username & password; there are a few [different configurations](https://www.home-assistant.io/docs/authentication/providers/).

> At least, let's please care having proper and working polyfills (also called "shims") always in place, in order to cover most current use cases and scenarios.

There is a polyfill that, until @leonelsr's comment, I believed covered the Google Chrome builtin manager and a few of the most popular password managers. This issue is for tracking the remaining password managers with which the polyfill does not work. I've edited the first comment in this chain to make this clearer, but I worry not everyone is reading it.

I hope we can keep discussion to the polyfill and how to improve it, since talking about re-writing the frontend with a separate set of components gets us nowhere.

**Comment by Nowaker** [comment]
bad bot

**Comment by leonelsr** [comment]
Chromium (Chrome and Edge, at least) built-in password manager still doesn't work.

**Comment by Nowaker** [comment]
> And the login page isn't always just a username & password; there are a few different configurations.

With usage rate of what, 5% at most? If any of these are enabled in that particular HA setup, sure, whatever, render the fancy dandy Web 3.0 rainbow. 

Otherwise, I'll repeat my words again: **Stop using fancy JavaScript shadow DOM / single page application / whatever comes after Web 2.0 frameworks to build a freaking login page. A simple HTML from the 90s will do better!**

**Comment by Nowaker** [comment]
bad bot

**Comment by Nowaker** [comment]
172 upvotes on the previous ticket: https://github.com/home-assistant/frontend/issues/3133 - we won't go away.

**Comment by kjake** [comment]
Ugh, this is annoying.  This doesn't work in Chrome, or the iOS companion app (Safari).

**Comment by steverep** [comment]
Adding `autocomplete` is my intended fix from an accessibility standpoint, which I suspect would fix the Safari/iOS issue.

As far as getting icons to show up and work for third-party browser extensions, I'd be inclined to throw the issue back to the extension if `autocomplete` doesn't work.

**Comment by RosemaryOrchard** [comment]
> Adding `autocomplete` is my intended fix from an accessibility standpoint, which I suspect would fix the Safari/iOS issue.

Just to confirm, this is correct. The [Apple Developer Documentation](https://developer.apple.com/documentation/security/password_autofill/enabling_password_autofill_on_an_html_input_element) shows it. Ideally,  it would be added to the 2FA token and password change fields too.

**Comment by steverep** [comment]
> Just to confirm, this is correct. The [Apple Developer Documentation](https://developer.apple.com/documentation/security/password_autofill/enabling_password_autofill_on_an_html_input_element) shows it. Ideally, it would be added to the 2FA token and password change fields too.

Yes, I can add it there and any other relevant places, just haven't gotten to it yet.

**Comment by steverep** [comment]
The `autocomplete` attribute has now been added in the following locations, and is in the current beta release for 2022.11.  At a minimum, I expect autofill on Safari/iOS to now work for all of these because it does not the polyfill for shadow DOM support.

- Login including MFA code
- Password changes and MFA setup
- HA cloud registration and login
- Creating new user during onboarding
- Many integrations that require external usernames/passwords during setup

However, please note that only the main login form currently uses the polyfill for shadow DOM, so browsers/extensions that still don't support shadow DOM won't see any effect in the other areas.  I know there are active bug reports on this for Chrome, Firefox, and popular extensions like BitWarden.

Please test with your particular password manager and report any unexpected results here.

**Comment by RosemaryOrchard** [comment]
Auto fill on iOS (specifically the iPad app) was broken today on 2022.11.0b6 as the 2FA code auto filled into the password field.

**Comment by Nowaker** [comment]
> However, please note that only the main login form currently uses the polyfill for shadow DOM, so browsers/extensions that still don't support shadow DOM won't see any effect in the other areas. I know there are active bug reports on this for Chrome, Firefox, and popular extensions like BitWarden.

We have an active bug report in Home Assistant:

**Please stop using fancy JavaScript shadow DOM / single page application / whatever comes after Web 2.0 frameworks to build a freaking login page? A simple HTML from the 90s will do better!**

**Comment by steverep** [comment]
@RosemaryOrchard can you inspect the DOM and verify the value of the `autocomplete` attribute is `"current-password"`?  I don't see how it could be anything else but I'll check to make sure I didn't mess up.  You are using a standard Home Assistant login correct?

If you can verify the value is correct, the culprit is probably whatever password manager you are using.

**Comment by RosemaryOrchard** [comment]
Sadly inspecting the DOM of the iOS app on an iPad is not something that is easy/possible to accomplish. I am using 1Password, but through the native iOS autofill. That said, it pops up Safari window which does not run any extensions or modifications, so it can't be one of those interfering with it.

**Comment by steverep** [comment]
You can use this extension to inspect on iOS:

https://apps.apple.com/us/app/web-inspector/id1584825745

**Comment by RosemaryOrchard** [comment]
You cannot use the inspection as it's an in app Safari, that means it uses [WKWebView](https://developer.apple.com/documentation/webkit/wkwebview). For clarity: You can't even use the native Safari keyboard shortcuts for autofill, extensions are not available (as stated above), and it doesn't even share cookies with Safari.

If I inspect in Safari itself, I can see the HTML which seems fine, but autofill does not work.
![8A4F87A3-EDE3-4C84-A4CA-6A9C68F7CBF1](https://user-images.githubusercontent.com/16113535/199372323-a89457b3-9c72-4c91-8bab-0a8d622d09ac.jpeg)
[ha_login.txt](https://github.com/home-assistant/frontend/files/9915050/ha_login.txt).

**Comment by steverep** [comment]
> You cannot use the inspection as it's an in app Safari, that means it uses WKWebView.

Right I was asking to inspect in Safari.  I would not expect different autofill behavior in the web view.

> Auto fill on iOS (specifically the iPad app) was broken today on 2022.11.0b6 as the 2FA code auto filled into the password field.

Are you saying that the username filled correctly?  I'm confused now because the screenshot says nothing was auto-filled.  Is that an error from 1Password?

**Comment by steverep** [comment]
@RosemaryOrchard I did make one more change to make sure that the fields have a `name` attribute in addition to `autocomplete`.  This is in 2022.11.0.  Please let me know if that makes a difference.

FWIW, I tested all the locations I listed with BitWarden on iOS and autofills performed as expected.

If anyone uses iCloud Keychain, some feedback on how that behaves in 2022.11.0 would be appreciated (or any other specific password manager for that matter).

**Comment by steverep** [comment]
FYI - the [Chromium issue for autofilling inside shadow DOM](https://bugs.chromium.org/p/chromium/issues/detail?id=649162) was just marked fixed

**Comment by pomeloy** [comment]
Auto fill is still broken on HA Frontend 20221213.0 with Safari 15.5 on macOS and Safari on iOS 16.1.1.

**Comment by steverep** [comment]
> Auto fill is still broken on HA Frontend 20221213.0 with Safari 15.5 on macOS and Safari on iOS 16.1.1.

With what password manager and on which page?

**Comment by pomeloy** [comment]
Using the default iCloud keychain on page `homeassistant:8123/auth/authorize?response_type=code&redirect_uri=http://homeassistant:8123/?auth_callback=1&client_id=http://homeassistant:8123/&state=...`

**Comment by steverep** [comment]
@pomeloy please explain your symptoms and steps on iOS because I cannot reproduce any issue.  Are you able to invoke autofill from the keyboard popup?  Or is it just not autofilling after you invoke it?

**Comment by pomeloy** [comment]
Actually there is no popup or autofill sign at all neither on various iOS devices nor on macOS so I am not able to invoke any autofill mechanism.

**Comment by iBobik** [comment]
I can confirm no password manager started on homepage. Having iCloud Keychain and Bitwarden. Just used it to login on GitHub, so I am sure it works. iOS 16.1.1
![259CC8A1-0316-43A1-89FC-5BE4DA4EC72C](https://user-images.githubusercontent.com/614232/208197635-4fc47c11-d37e-46ae-b1ff-c0735d3e45a2.png)

**Comment by steverep** [comment]
Thanks for the verification @iBobik.  It seems I'm making the mistake of testing with Firefox on iOS instead of Safari thinking the results would be the same as they both use webkit, but it only works on the former.

**Comment by m33x** [comment]
There is progress on Safari's side too. https://hachyderm.io/@rmondello/109752442226984975

**Comment by steverep** [comment]
For anyone using bitWarden, they just closed their issue regarding shadow DOM as completed (https://github.com/bitwarden/clients/issues/713).

**Comment by Coder84619** [comment]
I'm having this issue with 1Password on Mac. Using Safari (Ventura) and Vivaldi for Mac. Neither browser let 1Password auto-fill credentials.

**Comment by Coder84619** [comment]
@m33x The missing 1P icon is exactly the problem. On literally 99.9999% of other sites it shows up.

**Comment by SFoskett** [comment]
Still having issues in 1Password for Safari on iOS. Same password works perfectly on Safari for Mac.

**Comment by Eric13246** [comment]
I am experiencing this today on Firefox with LastPass on UI version 20230802.0. I have found a blog post that shows a possible solution to this problem here https://www.thisdot.co/blog/a-tale-of-form-autofill-litelement-and-the-shadow-dom/

I feel this would be a good compromise to render the login form only in the 'light DOM' while keeping the rest as web components contained in shadow DOMs.

**Comment by revolter** [comment]
Workaround for username and password on the Apple side: add and delete some random letter in both fields.

Unfortunately, the 2FA code autofill doesn't seem to have any workaround.

**Comment by steverep** [comment]
> I feel this would be a good compromise to render the login form only in the 'light DOM' while keeping the rest as web components contained in shadow DOMs.

That's been suggested as part of this issue before and my 2 🪙 as a volunteer developer is that's much easier said than done.  And at least when I think about it, the motivation factor is very low.  Maybe putting that in words would help...

First, the base text field component comes from an external package with complicated stylesheets, so it would have to be hacked to move it and errors would almost certainly arise.  Plus even on the login form, it's not just one shadow DOM but several nested ones, so you'd have to deal with potential conflicts from several components up the tree.  The alternative would be to use much simpler fields, which requires rewriting how the values are grabbed and sent to core for authentication, and that results in an a11y and UX trade off due to a lack of UI consistency.  A hybrid approach is already in place where there are hidden fields in the light DOM designed to catch autocompleted values and pass them to the visible web components, but it only seems to work for some password managers and not others.

Now lets say I spend the time to get past all that and it's approved for production, I've still only developed something that works on a couple fields on one page and not on the many others throughout the app that also have the potential to be autocompleted.  Moreover, the eventual goal would be to revert my entire set of changes when password managers, most of which have paid developers, get around to supporting fields within shadow DOMs, which have been around as part of the spec for many years now.

Continuing that last point, there also has been recent activity from sever managers, making this less and less an impactful thing to work on:

- bitWarden has a fixed release from earlier this year (https://github.com/bitwarden/clients/issues/713)
- Chrome has fixed their issue but looks to be working through their alpha/beta process ([Issue 649162: Related inputs separated by shadow root do not autofill](https://bugs.chromium.org/p/chromium/issues/detail?id=649162))
- Safari seems to have possibly fixed in 16.6 ([Bug 172567 - Safari autofill doesn't work in form inputs that are in a shadow root](https://bugs.webkit.org/show_bug.cgi?id=172567))
- Firefox has a bug report but no recent activity ([Bug 1629226 - [meta] Make password manager work with Shadow DOM](https://bugzilla.mozilla.org/show_bug.cgi?id=1629226)).  I would encourage folks to add their vote and/or comment on it.

I hear folks mentioning LastPass and 1Password not working, and from a quick search of their community forums, support has been requested but not implemented.  If you're a user of either, especially a paid one, I'd certainly be putting continued pressure on them.

**Comment by SFoskett** [comment]
I very much appreciate your answer, as it provides much needed perspective from a volunteer developer. That being said, it’s a little bit puzzling to me as a user that I don’t encounter this issue with any other systems. I believe you when you say that the issue is with the password software not the web application. But why don’t other systems have the same problem? And why has this not been a high priority for the two major password managers to fix? That’s really a rhetorical question, though I am curious.

I am a paid 1Password subscriber, and in fact, am an enterprise user, and will definitely submit this to them to fix. What specifically should I be asking for? “Shadow DOM support?” should I refer them to this thread? I am not (this kind of) developer, but I want to help.

Thank you for the time and energy you have invested into improving home assistant! And thank you for this answer.

**Comment by Roos-AID** [comment]
Steve,

the fix mentioned for Safari 16.6 does not solve the problem for the Apple password manager.

When the Apple password manager has autofilled the fields, following error is thrown :
Login attempt or request with invalid authentication from ipxxxxx (ipxxxx). Requested URL: '/auth/login_flow/144acb813613299cd87799xxxxxxxxxx'. (Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.6 Safari/605.1.15)

The only workaround is to have it autofilled,  manually ad a space (or any other char)  at end of userid and pw and remove this.
The UI then sees the value, without this the autofill string is not transferred to the core for authentication. 


Met vriendelijk groet,


Rob Roos

**Comment by steverep** [comment]
> I am a paid 1Password subscriber, and in fact, am an enterprise user, and will definitely submit this to them to fix. What specifically should I be asking for? “Shadow DOM support?” should I refer them to this thread? I am not (this kind of) developer, but I want to help.

@SFoskett  yes basically you would be asking for support for autofilling fields within shadow DOMs that are properly marked up with the `autocomplete` attribute.  And I would definitely point them to the links I provided showing their competitors actively working on it.



> Steve, the fix mentioned for Safari 16.6 does not solve the problem for the Apple password manager. When the Apple password manager has autofilled the fields, following error is thrown : Login attempt or request with invalid authentication from ...

@Roos-AID that's a separate issue likely caused by the hybrid approach I mentioned.  See #16976.

**Comment by Nowaker** [comment]
@steverep 
> First, the base text field component comes from an external package with complicated stylesheets, so it would have to be hacked to move it and errors would almost certainly arise. Plus even on the login form, it's not just one shadow DOM but several nested ones, so you'd have to deal with potential conflicts from several components up the tree. The alternative would be to use much simpler fields, which requires rewriting how the values are grabbed and sent to core for authentication, and that results in an a11y and UX trade off due to a lack of UI consistency. A hybrid approach is already in place where there are hidden fields in the light DOM designed to catch autocompleted values and pass them to the visible web components, but it only seems to work for some password managers and not others.

I'm going to repeat the SOLUTION again: Stop using fancy JavaScript shadow DOM / single page application / whatever comes after Web 2.0 frameworks to build a freaking login page. A simple HTML from the 90s will do better!

**Comment by steverep** [comment]
> I'm going to repeat the SOLUTION again: Stop using fancy JavaScript shadow DOM / single page application / whatever comes after Web 2.0 frameworks to build a freaking login page. A simple HTML from the 90s will do better!

@Nowaker thank you for pasting that yet again.  Since you directed it at me, I guess I'll reply once and only once.  The comment is deconstructive, disrespectful, combative, rude, etc.  It was so the 1st time and hasn't aged well either.  And frankly, it demonstrates just how little effort you're willing to put in to help.  Since you have the solution, please let everyone know when your non-fancy pull request is ready.

**Comment by Nowaker** [comment]
> The comment is deconstructive, disrespectful, combative, rude, etc.

It's all of those things that you mentioned because the leadership of this project fails to see the problem is in using fancy overengineered technologies for a login page.

I posted this comment two years ago: https://github.com/home-assistant/frontend/issues/10274#issuecomment-985699015 - in summary, it points out password-fill was a TOP1 requested feature, and that comments about this attracted tens of upvotes, all for nothing.

It's been going for 4 years now with no resolution so at this point, **repeating this is the only thing that's left to say** because there's nothing else left to say:

>> Stop using fancy JavaScript shadow DOM / single page application / whatever comes after Web 2.0 frameworks to build a freaking login page. A simple HTML from the 90s will do better!

**Comment by raman325** [comment]
> **repeating this is the only thing that's left to say** because there's nothing else left to say:

right. And saying it again actually detracts from the issue because it drowns out other potentially relevant comments.

Comments are going to start being marked as off topic or abusive if this language and behavior continues. Thanks.

**Comment by Coder84619** [comment]
I don't care what the solution is, but literally out of the hundreds of login forms I use with 1Password, HA is one maybe 2-3 that are broken. This really needs to be addressed by the developers.

**Comment by Roos-AID** [comment]
@steverep Steve, I referred to your remark that it is (possibly) being fixed by Safari developers : 

> Safari seems to have possibly fixed in 16.6 ([Bug 172567 - Safari autofill doesn't work in form inputs that are in a shadow root](https://bugs.webkit.org/show_bug.cgi?id=172567))
which is clearly not the case. 

Home Assistant is for me the only web page that does not work with Safari autofill.  

But I understand the complexity and challenges you (and all others) are facing to get this fixed.

**Comment by Nowaker** [comment]
> If you want to express your support or desire for this to be fixed, please +1 the original post.

@raman325 The original post with 172 upvotes - https://github.com/home-assistant/frontend/issues/3133 - is closed as "fixed" (haha) and locked. So nah, can't do that even if I could.

> Comments are going to start being marked as off topic or abusive if this language and behavior continues.

You can do what you please.

**Comment by chrgraham** [comment]
First-time new user, installed the latest official Docker container, opened the UI in the latest Chrome, was prompted to create an account and.... no `1Password` icons in the input fields. 

I thought I'd be writing an amazingly fresh bug report and instead found this ancient, crusty ticket from 2021.

**Comment by Nowaker** [comment]
@chrgraham From 2019 but in a new ticket... The original one is https://github.com/home-assistant/frontend/issues/3133 with 172 votes.

**Comment by raman325** [comment]
> @raman325 The original post with 172 upvotes - #3133 - is closed as "fixed" (haha) and locked. So nah, can't do that even if I could.

great, instead of the attitude, maybe you could have just asked for it to be unlocked? I'll do that in a sec.

> You can do what you please.

done, I can't ban you from posting here (I don't think), but my tolerance for your attitude is next to 0 at this point considering that I gave you a very straight forward response on what you could actually do to actually provide value to this conversation

**Comment by Coder84619** [comment]
> First-time new user, installed the latest official Docker container, opened the UI in the latest Chrome, was prompted to create an account and.... no `1Password` icons in the input fields. Also fails to prompt `1Password` to save the login, so you're off to reset your password first time you try to login again.
> 
> I thought I'd be writing an amazingly fresh bug report and instead found this ancient, crusty ticket from 2021.

I know..it's shocking the developers seem to not care about this issue. HA is about THE only site out of hundreds managed by 1Password that is broken. I simply don't understand the attitude of ignoring this issue. Pretty bad.

**Comment by sharkwhal** [comment]
> It does seem like the problem lies more in the hands of password manager vendors than in HA's hands.

So, in all seriousness and with all due respect, is the official direction “Let’s wait another 5 years and see what happens?”

At what point does it become worthwhile to accept the reality of the present and prioritize a solution? Surely many of the other recent UX improvements have been much more complex than this?

**Comment by iBobik** [comment]
As a developer who worked on a login form I confirm it is pain in the ass to make this work.

But we does not need to wait another 5 years - all major browsers support passkeys and we can skip login form thanks to it.

Let's focus on this.

**Comment by Eric13246** [comment]
@sharkwhal I understand the frustration it can cause. It's true that waiting for a big company to do something specific is wishful thinking, but at the same time we should not compromise code quality for something that would take long to develop and not be reusable somewhere else in the project. Maybe it's just an opportunity to ditch aging password managers? Who knows!

> @iBobik But we does not need to wait another 5 years - all major browsers support passkeys and we can skip login form thanks to it.

Yes please! I did not notice if HA's admin panel has an option for passwordless login or FIDO key support, but I agree this could be a better avenue.

**Comment by Nowaker** [comment]
> So, in all seriousness and with all due respect, is the official direction “Let’s wait another 5 years and see what happens?”

Mentioning passwordless, a new trend adopted by a tiny tiny minority of websites and applications, as a solution to HA's own technical shortcomings is exactly that, @sharkwhal. Thanks for saying that.

**Comment by sharkwhal** [comment]
I hope this doesn’t come across as argumentative, because that’s not my intent at all; I’m just trying to understand the decisions.

I’m not following the concern about reusability elsewhere in the project. There’s literally only one place I’d expect a password manager to work and that’s the login page. Not to mention this is the one part of the project that might be seen every time a user accesses HA, and is **the** first impression a new user will get. 

You’ve already seen a comment above from a first time user whose first impression of HA was this login page issue. There are countless other forum & Reddit posts along with the many upvotes on prior GitHub issues. 

Isn’t there a point where the poor UX and consistent user feedback justifies the time needed to develop a quality solution?

**Comment by sharkwhal** [comment]
@chrgraham - there is also [this topic in their forum](https://1password.community/discussion/139067/home-assistant-login-page-not-showing-autofill-button) 
It seems they have an internal issue filed, but I’d imagine it’s quite low in priority. However I’m sure if it was _their_ login page affected that would be a different story.

**Comment by steverep** [comment]
> I’m not following the concern about reusability elsewhere in the project. There’s literally only one place I’d expect a password manager to work and that’s the login page. Not to mention this is the one part of the project that might be seen every time a user accesses HA, and is **the** first impression a new user will get.

Fair question - please see my [comment above after implementing the `autocomplete` attribute](https://github.com/home-assistant/frontend/issues/10274#issuecomment-1295262991) for other places autofill should work.  Yes, the login form would certainly get the most utilization, but it should also work for password changes, two-factor setup, and any of the hundreds of integrations (or any blueprints) that need usernames, passwords, API keys, emails, etc. as part of their setup.  They can do that via the [text selector](https://www.home-assistant.io/docs/blueprint/selectors/#text-selector).

**Comment by steverep** [comment]
> @chrgraham - there is also [this topic in their forum](https://1password.community/discussion/139067/home-assistant-login-page-not-showing-autofill-button) 

That post is extremely helpful.  It makes it clear that 1password actually does work with the polyfill (i.e. hidden light DOM) in place, but the crux of the problem is that their injected buttons are not on top.  They are being overlapped by the "real" shadow DOM elements.

I wonder if they could fix the problem by simply adding a high [CSS `z-index`](https://www.home-assistant.io/docs/blueprint/selectors/#text-selector) to their styles.  Any 1password user could probably test this by finding their injected DOM in the browser inspector, and adding an attribute like `style="z-index: 99;"`.either editing the CSS or

**Comment by sharkwhal** [comment]
> I wonder if they could fix the problem by simply adding a high [CSS `z-index`](https://www.home-assistant.io/docs/blueprint/selectors/#text-selector) to their styles. Any 1password user could probably test this by finding their injected DOM in the browser inspector, and adding an attribute like `style="z-index: 99;"`.

Looks like this works:
<img width="977" alt="Screenshot 2023-08-30 at 7 47 03 PM" src="https://github.com/home-assistant/frontend/assets/53710689/8e5e5f76-08e9-4fab-a1c6-5fec357b80d5">

**Comment by steverep** [comment]
@sharkwhal could you please paste that DOM as text?  Specifically, I want to see exactly where 1password is injecting its content.

If anyone could do the same for LastPass, that would help as well.

**Comment by gjohansson-ST** [comment]
> If anyone could do the same for LastPass, that would help as well.

Did a test with lastpass and it doesn't seem to inject anything anywhere.
HA: 
![image](https://github.com/home-assistant/frontend/assets/62932417/21dbc04d-1aa3-41e8-ae2c-337737b40cf1)
Working example (github): 
![image](https://github.com/home-assistant/frontend/assets/62932417/75ea2e0d-c74d-4ed6-9a62-aad40ce942fb)

**Comment by steverep** [comment]
Thanks @gjohansson-ST.  I'm confused by that though because the OP says it does detect the polyfill but things are hidden similar to 1password.  @rianadon can you clarify here.  Or maybe their approach has changed?

For GH, what is the parent element for their injected `div`?

**Comment by gjohansson-ST** [comment]
They inject `<div data-lastpass-root="" style="position: absolute !important; top: 0px !important; left: 0px !important; height: 0px !important; width: 0px !important;"><div data-lastpass-infield="true" style="position: absolute !important; top: 0px !important; left: 0px !important;"></div></div>`directly in body.
Then `<div data-lastpass-icon-root="true" style="position: relative !important; height: 0px !important; width: 0px !important; float: left !important;"></div>` into the form. The latter seems to control the icons LP uses for fetching their info.

Attached is the github blank form loaded for lastpass input
[lp github.txt](https://github.com/home-assistant/frontend/files/12506980/lp.github.txt)

**Comment by steverep** [comment]
I made some changes to the polyfill to improve compatibility based on my testing with both LastPass and 1Password.  It should be in 2023.10.0.

I tested their extensions on Windows in both Chrome and Firefox.  Most of the changes were to appease LastPass, which was refusing to recognize the polyfill fields (for some very illogical and unpublished reasons).  The technical stuff is in https://github.com/home-assistant/frontend/pull/17830#issuecomment-1714329110.

➕ The good news is that LastPass will now autofill when invoked from the context or extension menus (1Password already does this fine).  Additionally, I seem to have some better success getting 1Password's icon to show up by tabbing past the login button, but it's quite finicky.

➖ The bad news is that LastPass' extension is putting its icon off screen so that's unusable (again see link above for technical details).  While LastPass seems to render its icons whenever the form is recognized, 1Password's icon seems to mainly be triggered by focusing one of the fields.  Since the polyfill form never receives focus, that's a problem.

I know folks are also having issues with Safari/iOS, which I did not test.  And honestly, it's not clear to me whether the password managers, Safari, or both are the prime factors.  Based on the history of comments in this issue, I'd guess Safari is the bigger factor as some success is commented right around the time Safari was adding support for shadow roots.

Anyway, I'd be remiss for not lastly mentioning some irony: neither LastPass nor 1Password look for input fields in shadow roots, yet both of their extensions inject menus and fields in shadow roots... 🙃

**Comment by piitaya** [comment]
We succeeded to remove shadow dom from the auth page (https://github.com/home-assistant/frontend/pull/18015).
So, with 2023.10 release... password managers should be fully supported! 🎉

I tested with 1Password and multiple browsers. The feature is in beta today so you can test it with other password managers 🙂

**Comment by Coder84619** [comment]
I'll wait until the GA release, but THANK YOU!!!!!!!!!!!

**Comment by Nowaker** [comment]
Beautiful, @piitaya! Finally, pragmatism prevailed! That's a big surprise - I really didn't expect it!

**Comment by raman325** [comment]
I believe this can be closed now. If anyone has any remaining issues, it's probably best to open a new thread
